pennylane/labs/resource_estimation/resource_utils.py

```diff
@@ -0,0 +1,281 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions for resource estimation."""
+from collections.abc import Callable
+from typing import Any
+
+import numpy as np
+
+
+# pylint: disable=unnecessary-lambda-assignment, too-many-arguments, not-callable
+def approx_poly_degree(
+    x_vec: np.ndarray,
+    target_func: Callable,
+    error_tol: float = 1e-6,
+    degrees: tuple[int, int] | int | None = None,
+    basis: str | None = None,
+    loss_func: Callable | None = None,
+    fit_func: Callable | None = None,
+    project_func: Callable | None = None,
+    **fit_kwargs: dict[str, Any],
+):
+    r"""Compute the minimum degree of a polynomial that fits the data with a given error tolerance.
+
+    Args:
+        x_vec (np.ndarray): the domain values for sampling the target function ``target_func``.
+            The minimum length of ``x_vec`` is two and is expected to be sorted.
+        target_func (Callable): function to be approximated with a polynomial and has the
+            signature: ``f(x_vec: np.ndarray) -> np.ndarray``.
+        error_tol (float): tolerance for the target fitting error. Defaults to ``1e-6``.
+            Unless ``loss_func`` is provided, this is the least squares fit error.
+        degrees (Tuple[int, int] | int | None): tuple of minimum and maximum degrees to consider.
+            Defaults to ``None``, which means all degrees from ``2`` to ``len(x_vec) - 1``
+            will be considered. If an integer is provided, it will be used as the maximum degree.
+        basis (str): basis to use for the polynomial. Available options are ``"chebyshev"``,
+            ``"legendre"``, and ``"hermite"``. Defaults to ``None``, which assumes fitting
+            data to a polynomial in the monomial basis.
+        loss_func (str | Callable | None): loss function to use, where available options are
+            ``"mse"`` (mean squared error), ``"mae"`` (mean absolute error), ``"rmse"``
+            (root mean squared error), ``"linf"`` (maximum absolute error), or a custom loss
+            function with the signature: ``f(pred: np.ndarray, target: np.ndarray) -> float``.
+            Defaults to ``None``, which means the least squares fit error is used.
+        fit_func (Callable | None): function that apaproximately fits the polynomial and has the signature:
+            ``f(x_vec: np.ndarray, y_vec: np.ndarray, deg: int, **fit_kwargs) -> tuple[Callable, float]``.
+            It should return a callable for the fit polynomial along with the least squares error of the fit.
+            Defaults to ``None``, which means the NumPy polynomial fitting function corresponding
+            to the ``basis`` keyword argument will be used. Note that, providing a custom function
+            will override the ``basis`` keyword argument.
+        project_func (str | Callable | None): function to project the dense interval
```

Comment:
Description needs improvement.

Reply:
Tried to improve it a bit, let me know if it reads better now.